### DISCUSSION TITLE
General recommendations for the best performance

### DISCUSSION THREAD

**Discussion by szaimen** [nextcloud/all-in-one#1335] [Wiki]
# General recommendations for the best performance

Since AIO is pretty much as optimized as it gets, it is mostly an infrastructural question how fast your Nextcloud performs. In general: much helps much.

## General recommendations
- Use SSD storage for everything and no HDD storage and especially no SD-cards or network storage like NFS or SMB (however all are fine as backup storage)
- Run Docker and AIO not in a VM but directly on the server (e.g. Docker Desktop needs 4GB+ additional RAM in comparison to a normal Linux server running Docker).
- Deactivate all optional containers and Nextcloud apps that you do not really need
- For faster previews, it is recommended to [set up the preview generator app](https://github.com/nextcloud/all-in-one/discussions/2810) in order to pre-generate previews
- Do not use Cloudflare (Proxy or Tunnel) as it potentially slows down things by a lot and introduces a lot of problems: https://github.com/nextcloud/all-in-one#notes-on-cloudflare-proxytunnel

## Additional recommendations, based on Nextcloud's user count and enabled containers
- Have 1 vCore per active user 
  (plus the base system requirement of AIO depending on the enabled containers which is mentioned in the AIO interface)
- Have 1GB RAM per active user 
  (plus the base system requirement of AIO depending on the enabled containers which is mentioned in the AIO interface)

## Additional recommendations for hosting in home networks
- Have a fiber-optic connection to the outside with the highest available upload rate
- Have a good router that is capable of handling everything that you throw at it
- Connect your server via Gbit LAN-cable and not Wi-fi to the router
- Set up a local DNS-server that allows to automatically connect to the server in the home network via the same domain by using the internal ip-address of the server. A simple way to achieve that, is to
    - add the [pi-hole community container](https://github.com/nextcloud/all-in-one/tree/main/community-containers/pi-hole) and [let docker use it via the daemon.json file](https://dockerlabs.collabnix.com/intermediate/networking/Configuring_DNS.html)
    - For more advanced use cases see [this doc](https://github.com/nextcloud/all-in-one#how-can-i-access-nextcloud-locally).

**Comment by brunoalfred** [comment]
Is there a guide to explaining on how to run AIO in k8s cluster?

**Reply by szaimen** [reply]
No. AIO is not designed to run in clusters.

**Reply by szaimen** [reply]
However you might get it running with docker swarm and https://github.com/nextcloud/all-in-one/tree/main/manual-install. Though, I never tried that and it is thus unspported.

**Comment by joshtrichards** [comment]
Might be worth calling that out Docker Desktop running on Windows is going to add +4 GB to any memory requirements due to the extra overhead of running a VM on the underlying host (unlike when the host is Linux-based):

https://docs.docker.com/desktop/install/windows-install/#system-requirements

**Reply by szaimen** [reply]
I've added it like so: 
![image](https://github.com/nextcloud/all-in-one/assets/42591237/37fdd774-deae-44df-a019-3d03fcedd8aa)

**Comment by alexsapps** [comment]
What sort of performance could we expect if we're doing everything right?

I'm using Nextcloud AIO on AWS EC2 (t3.medium + SSD / gp2 volume) and loading the files app and any individual file takes around 5 seconds. Loading the dashboard with no widgets besides "recent activity" takes 5 - 15 seconds.

**Reply by szaimen** [reply]
Hi, please ask this in a new discussion.

**Comment by alexanderdd** [comment]
@szaimen can you move this to the main readme? maybe under a heading like "system requirements" or "hardware recommendations", maybe also add "improving performance"?

The main readme is where I would first look for system requirements. This info here took quite a while to find.

(sidenote: maybe we even need a link from https://docs.nextcloud.com/server/latest/admin_manual/installation/system_requirements.html to the AIO system requirements, because AIO has become [the official way to install Nextcloud](https://docs.nextcloud.com/server/latest/admin_manual/installation/source_installation.html))

**Reply by szaimen** [reply]
The system requirements are shown in the AIO interface: https://github.com/nextcloud/all-in-one/blob/12ec7dca01ce0babbc1010f38a722c8075f5331b/php/templates/containers.twig#L606
